Build fails with strict-aliasing violations

I tried to build with the following *FLAGS to optimize the build: `-flto=4 -Werror=odr -Werror=lto-type-mismatch -Werror=strict-aliasing`

The -Werror=* flags are important to detect cases where the compiler can try to optimize based on assuming UB cannot happen, and miscompile code that has UB in it. strict-aliasing issues are always bad but LTO can make them even worse.

I got this error:
```
2024-07-15 03:37:17,743 root INFO x86_64-pc-linux-gnu-gcc -fno-strict-overflow -Wsign-compare -DNDEBUG -pipe -march=native -fstack-protector-all -O2 -fdiagnostics-color=always -frecord-gcc-switches -flto=4 -Werror=odr -Werror=lto-type-mismatch -Werror=strict-aliasing -Wformat -Werror=format-security -Werror=implicit-function-declaration -Werror=implicit-int -Werror=int-conversion -Werror=incompatible-pointer-types -DNDEBUG -fPIC -I/usr/include/postgresql-16 -I/usr/include/python3.12 -c psycopg_c/_psycopg.c -o /var/tmp/portage/dev-python/psycopg-3.1.19/work/psycopg-3.1.19-python3_12/build/temp.linux-x86_64-cpython-312/psycopg_c/_psycopg.o
psycopg_c/_psycopg.c: In function ‘__pyx_f_9psycopg_c_8_psycopg__array_load_binary’:
psycopg_c/_psycopg.c:34509:16: warning: assignment discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
34509 | __pyx_v_dimptr = (__pyx_v_buf + (3 * (sizeof(uint32_t))));
      |                ^
In file included from /usr/include/sys/types.h:176,
                 from /usr/include/stdlib.h:514,
                 from /usr/include/python3.12/Python.h:23,
                 from psycopg_c/_psycopg.c:42:
psycopg_c/_psycopg.c: In function ‘__pyx_f_9psycopg_c_8_psycopg_20IntervalBinaryLoader_cload’:
psycopg_c/_psycopg.c:49642:52: error: dereferencing type-punned pointer will break strict-aliasing rules [-Werror=strict-aliasing]
49642 | __pyx_v_val = be64toh((((uint64_t *)__pyx_v_bedata)[0]));
      |                                                    ^
psycopg_c/_psycopg.c: In function ‘__pyx_f_9psycopg_c_8_psycopg_17FloatBinaryDumper_cdump’:
psycopg_c/_psycopg.c:56365:52: error: dereferencing type-punned pointer will break strict-aliasing rules [-Werror=strict-aliasing]
56365 | __pyx_v_beval = htobe64((((uint64_t *)(&__pyx_v_d))[0]));
      |                                                    ^
psycopg_c/_psycopg.c: In function ‘__pyx_f_9psycopg_c_8_psycopg_18Float4BinaryDumper_cdump’:
psycopg_c/_psycopg.c:56557:52: error: dereferencing type-punned pointer will break strict-aliasing rules [-Werror=strict-aliasing]
56557 | __pyx_v_beval = htobe32((((uint32_t *)(&__pyx_v_f))[0]));
      |                                                    ^
psycopg_c/_psycopg.c: In function ‘__pyx_f_9psycopg_c_8_psycopg_19NumericBinaryLoader_cload’:
psycopg_c/_psycopg.c:57961:20: warning: assignment discards ‘const’ qualifier from pointer target type [-Wdiscarded-qualifiers]
57961 |   __pyx_v_digitptr = (__pyx_v_data + (sizeof(__pyx_v_behead)));
      |                    ^
```

Reported downstream at https://bugs.gentoo.org/935401
Full build log: [build.log](https://github.com/user-attachments/files/16231403/build.log)

--- a/.github/workflows/tests.yml
+++ b/.github/workflows/tests.yml
@@ -126,7 +126,7 @@ jobs:
       - name: Configure memory alignment tests
         if: ${{ matrix.ext == 'align' }}
         run: |
-          echo "CFLAGS=-fsanitize=undefined" >> $GITHUB_ENV
+          echo "CFLAGS=-fsanitize=undefined -Werror=strict-aliasing -Werror=odr -Werror=lto-type-mismatch" >> $GITHUB_ENV
           echo "UBSAN_OPTIONS=halt_on_error=1" >> $GITHUB_ENV
           echo "PYTEST_ADDOPTS=-v" >> $GITHUB_ENV
 

--- a/psycopg_c/psycopg_c/pq/pgconn.pyx
+++ b/psycopg_c/psycopg_c/pq/pgconn.pyx
@@ -710,7 +710,7 @@ cdef int _call_int(PGconn pgconn, conn_int_f func) except -2:
     return func(pgconn._pgconn_ptr)
 
 
-cdef void notice_receiver(void *arg, const libpq.PGresult *res_ptr) with gil:
+cdef void notice_receiver(void *arg, const libpq.PGresult *res_ptr) noexcept with gil:
     cdef PGconn pgconn = <object>arg
     if pgconn.notice_handler is None:
         return

--- a/psycopg_c/psycopg_c/types/array.pyx
+++ b/psycopg_c/psycopg_c/types/array.pyx
@@ -232,7 +232,7 @@ cdef object _array_load_binary(
     cdef Py_ssize_t[MAXDIM] dims
     cdef int i
     cdef uint32_t bedata
-    cdef char *dimptr = buf + 3 * sizeof(uint32_t)
+    cdef const char *dimptr = buf + 3 * sizeof(uint32_t)
     for i in range(ndims):
         # Every dimension is dim, lower bound
         memcpy(&bedata, dimptr, sizeof(bedata))

--- a/psycopg_c/psycopg_c/types/datetime.pyx
+++ b/psycopg_c/psycopg_c/types/datetime.pyx
@@ -971,11 +971,11 @@ cdef class IntervalBinaryLoader(CLoader):
     format = PQ_BINARY
 
     cdef object cload(self, const char *data, size_t length):
-        cdef int32_t bedata[4]
+        cdef int64_t bedata[2]
         memcpy(&bedata, data, sizeof(bedata))
-        cdef int64_t val = endian.be64toh((<uint64_t *>bedata)[0])
-        cdef int32_t days = endian.be32toh(bedata[2])
-        cdef int32_t months = endian.be32toh(bedata[3])
+        cdef int64_t val = endian.be64toh(bedata[0])
+        cdef int32_t days = endian.be32toh((<uint32_t *>bedata)[2])
+        cdef int32_t months = endian.be32toh((<uint32_t *>bedata)[3])
 
         cdef int years
         with cython.cdivision(True):

--- a/psycopg_c/psycopg_c/types/numeric.pyx
+++ b/psycopg_c/psycopg_c/types/numeric.pyx
@@ -393,10 +393,9 @@ cdef class Float4BinaryLoader(CLoader):
         cdef uint32_t bedata
         memcpy(&bedata, data, sizeof(bedata))
         cdef uint32_t asint = endian.be32toh(bedata)
-        # avoid warning:
-        # dereferencing type-punned pointer will break strict-aliasing rules
-        cdef char *swp = <char *>&asint
-        return PyFloat_FromDouble((<float *>swp)[0])
+        cdef float f
+        memcpy(&f, &asint, sizeof(asint))
+        return PyFloat_FromDouble(f)
 
 
 @cython.final
@@ -408,8 +407,9 @@ cdef class Float8BinaryLoader(CLoader):
         cdef uint64_t bedata
         memcpy(&bedata, data, sizeof(bedata))
         cdef uint64_t asint = endian.be64toh(bedata)
-        cdef char *swp = <char *>&asint
-        return PyFloat_FromDouble((<double *>swp)[0])
+        cdef double d
+        memcpy(&d, &asint, sizeof(asint))
+        return PyFloat_FromDouble(d)
 
 
 @cython.final

--- a/psycopg_c/psycopg_c/types/numeric.pyx
+++ b/psycopg_c/psycopg_c/types/numeric.pyx
@@ -471,7 +471,7 @@ cdef class NumericBinaryLoader(CLoader):
         cdef int i
         cdef PyObject *pctx
         cdef object key
-        cdef char *digitptr
+        cdef const char *digitptr
         cdef uint16_t bedigit
 
         if sign == NUMERIC_POS or sign == NUMERIC_NEG:

--- a/psycopg_c/psycopg_c/types/numeric.pyx
+++ b/psycopg_c/psycopg_c/types/numeric.pyx
@@ -346,7 +346,9 @@ cdef class FloatBinaryDumper(CDumper):
 
     cdef Py_ssize_t cdump(self, obj, bytearray rv, Py_ssize_t offset) except -1:
         cdef double d = PyFloat_AsDouble(obj)
-        cdef uint64_t beval = endian.htobe64((<uint64_t *>&d)[0])
+        cdef uint64_t ival
+        memcpy(&ival, &d, sizeof(ival))
+        cdef uint64_t beval = endian.htobe64(ival)
         cdef uint64_t *buf = <uint64_t *>CDumper.ensure_size(
             rv, offset, sizeof(beval))
         memcpy(buf, &beval, sizeof(beval))
@@ -361,7 +363,9 @@ cdef class Float4BinaryDumper(CDumper):
 
     cdef Py_ssize_t cdump(self, obj, bytearray rv, Py_ssize_t offset) except -1:
         cdef float f = <float>PyFloat_AsDouble(obj)
-        cdef uint32_t beval = endian.htobe32((<uint32_t *>&f)[0])
+        cdef uint32_t ival
+        memcpy(&ival, &f, sizeof(ival))
+        cdef uint32_t beval = endian.htobe32(ival)
         cdef uint32_t *buf = <uint32_t *>CDumper.ensure_size(
             rv, offset, sizeof(beval))
         memcpy(buf, &beval, sizeof(beval))

